sdk/metric/aggregator/exponential/structure/exponential.go

```diff
@@ -0,0 +1,657 @@
+	}
+	newScale := h.mapping.Scale() - change
+
+	h.positive.downscale(change)
+	h.negative.downscale(change)
+	var err error
+	h.mapping, err = newMapping(newScale)
+	if err != nil {
+		panic(fmt.Sprint("impossible scale", newScale))
+	}
+}
+
+// changeScale computes how much downscaling is needed by shifting the
+// high and low values until they are separated by no more than size.
+func changeScale(hl highLow, size int32) int32 {
+	var change int32
+	for hl.high-hl.low >= size {
+		hl.high >>= 1
+		hl.low >>= 1
+		change++
+	}
+	return change
+}
+
+// update increments the appropriate buckets for a given absolute
+// value by the provided increment.
+func (h *Histogram[N]) update(b *Buckets, value float64, incr uint64) {
+	index := h.mapping.MapToIndex(value)
+
+	hl, success := h.incrementIndexBy(b, index, incr)
+	if success {
+		return
+	}
+
+	h.downscale(changeScale(hl, h.maxSize))
+
+	index = h.mapping.MapToIndex(value)
+	if _, success := h.incrementIndexBy(b, index, incr); !success {
+		panic("downscale logic error")
+	}
+}
+
+// incrementIndexBy determines if the index lies inside the current range
+// [indexStart, indexEnd] and, if not, returns the minimum size (up to
+// maxSize) will satisfy the new value.
+func (h *Histogram[N]) incrementIndexBy(b *Buckets, index int32, incr uint64) (highLow, bool) {
+	if incr == 0 {
+		// Skipping a bunch of work for 0 increment.  This
+		// happens when merging sparse data, for example.
+		// This also happens UpdateByIncr is used with a 0
+		// increment, means it can be safely skipped.
+		return highLow{}, true
+	}
+	if b.Len() == 0 {
+		if b.backing == nil {
+			b.backing = &bucketsVarwidth[uint8]{
+				counts: []uint8{0},
+			}
+		}
+		b.indexStart = index
+		b.indexEnd = b.indexStart
+		b.indexBase = b.indexStart
+	} else if index < b.indexStart {
+		if span := b.indexEnd - index; span >= h.maxSize {
+			// rescale needed: mapped value to the right
+			return highLow{
+				low:  index,
+				high: b.indexEnd,
+			}, false
+		} else if span >= b.backing.size() {
+			h.grow(b, span+1)
+		}
+		b.indexStart = index
+	} else if index > b.indexEnd {
+		if span := index - b.indexStart; span >= h.maxSize {
+			// rescale needed: mapped value to the left
+			return highLow{
+				low:  b.indexStart,
+				high: index,
+			}, false
+		} else if span >= b.backing.size() {
+			h.grow(b, span+1)
+		}
+		b.indexEnd = index
+	}
+
+	bucketIndex := index - b.indexBase
+	if bucketIndex < 0 {
+		bucketIndex += b.backing.size()
+	}
+	b.incrementBucket(bucketIndex, incr)
+	return highLow{}, true
+}
+
+// grow resizes the backing array by doubling in size up to maxSize.
+// this extends the array with a bunch of zeros and copies the
+// existing counts to the same position.
+func (h *Histogram[N]) grow(b *Buckets, needed int32) {
+	size := b.backing.size()
+	bias := b.indexBase - b.indexStart
+	oldPositiveLimit := size - bias
+	newSize := int32(1) << (32 - bits.LeadingZeros32(uint32(needed)))
```

Comment:
why use bit shifting here instead of `2*needed`? Given the MaximumMaxSize is defined to be 16384, multiplying this by 2 is not going to overflow.

Reply:
Nevermind the bucket can be []uint8

Reply:
Would be great to have a comment here explaining why shifting bits here is necessary

Reply:
Thanks for asking. This expression was computing a rounded power-of-two, but I noticed it's rounding up unnecessarily when the requested size is already a power of two. Leaving this comment open while I resolve that. See https://stackoverflow.com/questions/466204/rounding-up-to-next-power-of-2 there are many approaches.

You're right that this could just double the required size -- I was trying to align with powers of two to help the memory allocator, which I still think is worthwhile. Shifting bits may seem like overkill in this case, however there are at least 7 other bit-shift expressions in this file, so it's not really out of place. 

After reading the above stackoverflow, I really like this:
```
	v--
	v |= v >> 1
	v |= v >> 2
	v |= v >> 4
	v |= v >> 8
	v |= v >> 16
	v++
```
though it's way less obvious why that works. 😀

Reply:
I replaced the original expression with the math/bits-free variation above and comments indicating where it comes from. I also left in the comment the former expression based on math/bits, corrected with a -1. Any of these forms works, and since the code is designed to re-use memory after it's allocated, there's no logic that would have caught the former off-by-one (which could result in twice as much memory as needed being allocated) and no test was impacted.